File: CS/PersistentRowState/Classes.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Windows;
using System.Windows.Controls;
using System.Windows.Controls.Primitives;
using System.Windows.Data;
using DevExpress.Mvvm.UI.Interactivity;
using DevExpress.Xpf.Core;
using DevExpress.Xpf.Core.Native;
using DevExpress.Xpf.Grid;
using DevExpress.Xpf.Utils;

namespace PersistentRowState
{
    public class ResizableDataRow : Control , IResizeHelperOwner
    {
        public static readonly DependencyProperty RowHeightProperty = DependencyProperty.RegisterAttached(
            "RowHeight",
            typeof(double), typeof(ResizableDataRow),
            new PropertyMetadata(20d));

        static ResizableDataRow()
        {
            RowData.RowDataProperty.OverrideMetadata(typeof(ResizableDataRow), new FrameworkPropertyMetadata(OnScrollViewerVerticalOffsetChanged));
        }

        static void OnScrollViewerVerticalOffsetChanged(DependencyObject d, DependencyPropertyChangedEventArgs e)
        {
            ((ResizableDataRow)d).OnRowDataChanged((RowData)e.OldValue, (RowData)e.NewValue);
        }

        public static void SetRowHeight(DependencyObject element, double value)
        {
            element.SetValue(RowHeightProperty, value);
        }

        public static double GetRowHeight(DependencyObject element)
        {
            return (double)element.GetValue(RowHeightProperty);
        }

        ResizeHelper resizeHelper;

        RowData RowData
        {
            get { return (RowData)DataContext; }
        }

        double RowHeight
        {
            get { return GetRowHeight(RowData.RowState); }
            set { SetRowHeight(RowData.RowState, value); }
        }

        public override void OnApplyTemplate()
        {
            resizeHelper = new ResizeHelper(this);
            base.OnApplyTemplate();
            Thumb resizer = (Thumb)GetTemplateChild("PART_Resizer");
            resizeHelper.Init(resizer);
            InitializeRowHeight();
        }
        void OnRowDataChanged(RowData oldValue, RowData newValue)
        {
            if (oldValue != null)
                oldValue.ContentChanged -= new EventHandler(RowData_ContentChanged);
            if (newValue != null)
            {
                newValue.ContentChanged += new EventHandler(RowData_ContentChanged);
                InitializeRowHeight();
            }
        }
        void RowData_ContentChanged(object sender, EventArgs e)
        {
            InitializeRowHeight();
        }
        void InitializeRowHeight()
        {
            if (RowHeight == 0)
                RowHeight = 20;
        }

        #region IResizeHelperOwner Members
        double IResizeHelperOwner.ActualSize { get { return RowHeight; } set { RowHeight = value; } }
        void IResizeHelperOwner.ChangeSize(double delta)
        {
            RowHeight = Math.Min(300, Math.Max(20, RowHeight + delta));
        }
        void IResizeHelperOwner.OnDoubleClick() { }
        void IResizeHelperOwner.SetIsResizing(bool isResizing) { }
        SizeHelperBase IResizeHelperOwner.SizeHelper { get { return VerticalSizeHelper.Instance; } }
        #endregion
    }

}

File: VB/PersistentRowState/Classes.vb
Imports System
Imports System.Windows
Imports System.Windows.Controls
Imports System.Windows.Controls.Primitives
Imports DevExpress.Xpf.Core
Imports DevExpress.Xpf.Grid

Namespace PersistentRowState

    Public Class ResizableDataRow
        Inherits Control
        Implements IResizeHelperOwner

        Public Shared ReadOnly RowHeightProperty As DependencyProperty = DependencyProperty.RegisterAttached("RowHeight", GetType(Double), GetType(ResizableDataRow), New PropertyMetadata(20R))

        Shared Sub New()
            Call RowData.RowDataProperty.OverrideMetadata(GetType(ResizableDataRow), New FrameworkPropertyMetadata(AddressOf OnScrollViewerVerticalOffsetChanged))
        End Sub

        Private Shared Sub OnScrollViewerVerticalOffsetChanged(ByVal d As DependencyObject, ByVal e As DependencyPropertyChangedEventArgs)
            CType(d, ResizableDataRow).OnRowDataChanged(CType(e.OldValue, RowData), CType(e.NewValue, RowData))
        End Sub

        Public Shared Sub SetRowHeight(ByVal element As DependencyObject, ByVal value As Double)
            element.SetValue(RowHeightProperty, value)
        End Sub

        Public Shared Function GetRowHeight(ByVal element As DependencyObject) As Double
            Return CDbl(element.GetValue(RowHeightProperty))
        End Function

        Private resizeHelper As ResizeHelper

        Private ReadOnly Property RowData As RowData
            Get
                Return CType(DataContext, RowData)
            End Get
        End Property

        Private Property RowHeight As Double
            Get
                Return GetRowHeight(RowData.RowState)
            End Get

            Set(ByVal value As Double)
                SetRowHeight(RowData.RowState, value)
            End Set
        End Property

        Public Overrides Sub OnApplyTemplate()
            resizeHelper = New ResizeHelper(Me)
            MyBase.OnApplyTemplate()
            Dim resizer As Thumb = CType(GetTemplateChild("PART_Resizer"), Thumb)
            resizeHelper.Init(resizer)
            InitializeRowHeight()
        End Sub

        Private Sub OnRowDataChanged(ByVal oldValue As RowData, ByVal newValue As RowData)
            If oldValue IsNot Nothing Then RemoveHandler oldValue.ContentChanged, New EventHandler(AddressOf RowData_ContentChanged)
            If newValue IsNot Nothing Then
                AddHandler newValue.ContentChanged, New EventHandler(AddressOf RowData_ContentChanged)
                InitializeRowHeight()
            End If
        End Sub

        Private Sub RowData_ContentChanged(ByVal sender As Object, ByVal e As EventArgs)
            InitializeRowHeight()
        End Sub

        Private Sub InitializeRowHeight()
            If RowHeight = 0 Then RowHeight = 20
        End Sub

'#Region "IResizeHelperOwner Members"
        Private Property ActualSize As Double Implements IResizeHelperOwner.ActualSize
            Get
                Return RowHeight
            End Get

            Set(ByVal value As Double)
                RowHeight = value
            End Set
        End Property

        Private Sub ChangeSize(ByVal delta As Double) Implements IResizeHelperOwner.ChangeSize
            RowHeight = Math.Min(300, Math.Max(20, RowHeight + delta))
        End Sub

        Private Sub OnDoubleClick() Implements IResizeHelperOwner.OnDoubleClick
        End Sub

        Private Sub SetIsResizing(ByVal isResizing As Boolean) Implements IResizeHelperOwner.SetIsResizing
        End Sub

        Private ReadOnly Property SizeHelper As SizeHelperBase Implements IResizeHelperOwner.SizeHelper
            Get
                Return VerticalSizeHelper.Instance
            End Get
        End Property
'#End Region
    End Class
End Namespace
